doc/developer/design/20220411_reclocking_implementation.md

```diff
@@ -0,0 +1,257 @@
+* Describe how users interact with the reclocking data.
+
+## Non-Goals
+
+* Motivating reclocking. See the product spec in [#11218].
+* Describing reclocking formally. See the
+  [original reclocking design doc](20210714_reclocking.md).
+
+## Description
+
+### Reclockable source types
+
+We define a **reclockable** source type as one where the upstream system
+provides a useful [gauge of progress](20210714_reclocking.md#description).
+The following source types are reclockable:
+
+* File
+* Kafka
+* Kinesis
+* PostgreSQL
+* S3
+
+The following sources types are nonreclockable:
+
+* PubNub
+
+### Storage layer implementation
+
+A reclockable source will consist of two definite collections in the storage
+layer: the [`remap`](20210714_reclocking.md#remap) collection, which contains
+the reclocking metadata, and the reclocked data collection.
+
+Nonreclockable sources will consist of only a single definite collection
+containing the source's data. For consistency with reclockable sources, we will
+refer to this collection as the "reclocked" collection, even though there is no
+upstream clock.
+
+The type of the `remap` collection depends on the source type:
+
+| Source type | `remap` type                                     |
+|-------------|--------------------------------------------------|
+| File        | `(ignored: (), byte_offset_increment: i64)`      |
+| Kafka       | `(partition: i32, offset_increment: i64)`        |
+| Kinesis     | `(shard: i64, sequence_increment: numeric)`      |
+| PostgreSQL  | `(ignored: (), lsn_increment: i64)`              |
+| S3          | `(object_key: text, byte_offset_increment: i64)` |
+
+The `remap` collection exists in the Materialize time domain, where a `u64`
+indicates milliseconds since the Unix epoch.
+
+Let's take a concrete example. Suppose the `remap` collection for a two
+partition Kafka source contains the following entries:
+
+```
+# (remap,   time,           diff)
+  ((1, +42), 1649686076392,   +1)    # (1)
+  ((2, +40), 1649686076392,   +1)    # (2)
+  ((1, +3),  1649686079487,   +1)    # (3)
+  ((2, +4),  1649686079487,   +1)    # (4)
+```
+
+Entry (1) indicates that offsets 0-42 (inclusive) in partition 1 are assigned a
+Materialize timestamp of Monday, April 11, 2022 14:07:56.392 PM UTC. Entry (2)
+indicates that offsets 0-40 in partition 2 are assigned the same timestamp.
+Entry (3) indicates that offsets 43-45 in partition 1 are assigned a Materialize
+timestamp of 14:07:59.487 UTC on the same day. Entry (4) indicates that offsets
+41-44 in partition 2 are assigned the same timestamp.
+
+Notice how the offsets are described by their *increments*, not their absolute
+value. This allows the collection to be compacted using `partition` as the
+compaction key. This generalizes to the other reclockable source types.
+
+#### `remap` policy
+
+The mapping must always preserve the order of events from the upstream source
+whenever such an order is present. The mapping must also preserve the
```

Comment:
This seems to literally state that we cannot coalesce transactions. Is that the intent, or is it the weaker "coalesce, don't mis-order events"?

Reply:
Fixed!